packages/shylu/shylu_dd/bddc/src/shylu_DofManager.h

```diff
@@ -265,37 +267,39 @@ template <class LO, class GO>
     indices.erase(iter, indices.end());
   }
 
-  static void generateRootMap(RCP<const Map> & inputMap,
-                              int root,
-                              RCP<const Map> & rootMap)
+  static void generateRootMap(RCP<const Map> & inputMap, 
+			      int root,
+			      RCP<const Map> & rootMap)
   {
     // we assume InputMap has uniquely owned rows
     RCP<const Teuchos::Comm<int> > Comm = inputMap->getComm();
     int numProc = Comm->getSize();
     LO sendCount = inputMap->getNodeNumElements();
     std::vector<LO> recvCounts(numProc), displs(numProc);
     Teuchos::gather<int, LO> (&sendCount, 1, &recvCounts[0], 1,
-                              root, *Comm);
+			      root, *Comm);
     int numGIDs(0);
     std::vector<GO> myNewGIDs;
     if (Comm->getRank() == root) {
       for (int i=0; i<numProc; i++) {
-        numGIDs += recvCounts[i];
-        if (i < numProc-1) displs[i+1] = displs[i] + recvCounts[i];
+	numGIDs += recvCounts[i];
+	if (i < numProc-1) displs[i+1] = displs[i] + recvCounts[i];
       }
-      assert (numGIDs == LO(inputMap->getGlobalNumElements()));
+      const bool test = (numGIDs == LO(inputMap->getGlobalNumElements()));
+      BDDC_TEST_FOR_EXCEPTION(test == false, std::runtime_error, 
+			      "invalud numGIDs");
       myNewGIDs.resize(numGIDs);
     }
     const GO* globalIDs = (inputMap->getNodeElementList()).getRawPtr();
-    Teuchos::gatherv<int, LO> (globalIDs, sendCount, myNewGIDs.data(),
-                               recvCounts.data(), displs.data(), root, *Comm);
-    rootMap = rcp( new Map(-1, Teuchos::ArrayView<GO>(myNewGIDs), 0,
-                           Comm) );
+    Teuchos::gatherv<int, LO> (globalIDs, sendCount, myNewGIDs.data(), 
```

Comment:
`Teuchos::gatherv` doesn't necessarily do what you think it does, and doesn't actually do the same thing that `MPI_Gatherv` does.  See implementation of Tpetra's `computeGatherMap` function.

Reply:
Let's discuss this when you are available. Thank you for the heads up.

Reply:
Teuchos may be OK, but you can just call `Tpetra::Details::computeGatherMap`.

Reply:
See `packages/tpetra/core/src/Tpetra_ComputeGatherMap.hpp`.